keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,145 @@
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Startup
+An application that has a proxy container acting as a sidecar may fail when it starts up as it's unable to communicate until its proxy has started up successfully. Readiness probes don't help if the application is trying to talk outbound.
+
+### Shutdown
+Applications that rely on sidecars may experience a high amount of errors when shutting down as the sidecar may terminate before the application has finished what it's doing.
+
+
+## Goals
+
+Solve issues so that they don't require application modification:
+[25908](https://github.com/kubernetes/kubernetes/issues/25908) - Job completion
+[65502](https://github.com/kubernetes/kubernetes/issues/65502) - Container startup dependencies
+
+## Non-Goals
+
+Allowing multiple containers to run at once during the init phase.
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+Sidecars will be started before normal containers but after init, so that they are ready before your main processes start.
+
+This will change the Pod startup to look like this:
+* Init containers start
+* Init containers finish
+* Sidecars start
+* Containers start
+
+During pod termination sidecars will be terminated last:
+* Containers sent SIGTERM
+* Sidecars sent SIGTERM
+
+Both will still obey the grace period before being sent a SIGKILL
+
+PreStop Hooks will be sent to sidecars and containers at the same time.
+This will be useful in scenarios such as when your sidecar is a proxy so that it knows to no longer accept inbound requests but can continue to allow outbound ones until the the primary containers have shut down.
```

Comment:
In general we do not advise apps to immediately stop accepting connections - external LBs need time.

Reply:
@thockin So I take it the concern here would be that the loadbalancer takes a while to realise that the pod is no longer a valid target so continues to send traffic which will just fail as the pod is no longer accepting connections. I can see how that could cause issues.
This specific example was mentioned by some of the folks from istio, maybe @nmittler can elaborate, or maybe I misinterpreted their use case.

Reply:
That's right.  I know some apps interpret SIGTERM to mean to stop accepting traffic, but that is actively harmful in k8s, where everything is async.